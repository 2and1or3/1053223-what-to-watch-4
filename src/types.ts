interface FilmType {
  id: number;
  title: string;
  preview: string;
  background: string;
  backgroundColor: string;
  cover: string;
  genre: string;
  isFavorite: boolean;
  src: string;
  release: string;
  description: string;
  poster: string;
  duration: number;
  rating: number;
  voiceCount: number;
  director: string;
  actors: string[];
}

interface CommentType {
  id: string;
  author: string;
  date: string;
  description: string;
  rate: string;
}

interface GenreType {
  id: string;
  title: string;
}

type RefType = () => void | {current: React.ReactNode};

export {FilmType, CommentType, RefType, GenreType};
