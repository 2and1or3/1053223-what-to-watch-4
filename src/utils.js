import {AppRoute, MONTHS} from './consts';

const createVideoMock = (element) => {
  if (element.type === `video`) {
    return element;
  }

  return null;
};


const extend = (a, b) => {
  return Object.assign({}, a, b);
};

const adapterToLocalFilms = (films) => {

  const localFilms =
    films
    .slice()
    .map((film) => ({
      id: film.id.toString(),
      isFavorite: film.is_favorite,
      title: film.name,
      poster: film.preview_image,
      preview: film.preview_video_link,
      src: film.video_link,
      background: film.background_image,
      backgroundColor: film.background_color,
      cover: film.poster_image,
      genre: film.genre.toLowerCase(),
      release: film.released.toString(),
      description: film.description,
      rating: film.rating,
      voiceCount: film.scores_count,
      duration: film.run_time,
      director: film.director,
      actors: film.starring.slice(),
    }));

  return localFilms;
};

const adapterToLocalComments = (comments) => {

  const localComments =
  comments
  .map((comment) => ({
    id: comment.id.toString(),
    author: comment.user.name,
    date: comment.date,
    description: comment.comment,
    rate: comment.rating.toString(),
  }));

  return localComments;
};

const GetPath = {
  filmReview: (id) => `${AppRoute.FILM}${id}${AppRoute.REVIEW}`,
  filmPlayer: (id) => `${AppRoute.FILM}${id}${AppRoute.PLAYER}`,
  filmDetails: (id) => `${AppRoute.FILM}${id}`,
};

const getDateFormat = (date) => {
  const data = new Date(date);
  const month = data.getMonth();
  const day = data.getDate();
  const year = data.getFullYear();

  return `${MONTHS[month]} ${day} ${year}`;
};

const getWatchTimeFormat = (seconds) => {

  let hours = Math.floor(seconds / 3600);
  let min = Math.floor((seconds - hours * 3600) / 60);
  let sec = seconds - hours * 3600 - min * 60;

  hours = hours.toString();
  min = min.toString().padStart(2, `0`);
  sec = sec.toString().padStart(2, `0`);

  return `${hours}:${min}:${sec}`;
};

const capitalize = (str) => {
  str = str.toLowerCase();
  let words = str.split(` `);

  words = words.map((word) => word[0].toUpperCase() + word.slice(1));
  str = words.join(` `);

  return str;
};

const getUniqueGenres = (genres) => {
  let uniqueGenres = [];

  genres.forEach((genre) => {
    const isRepeated = uniqueGenres.some((uniqueGenre) => uniqueGenre === genre);

    if (!isRepeated) {
      uniqueGenres.push(genre);
    }
  });


  const localUniqueGenres = uniqueGenres.map((uniqueGenre) => ({
    id: uniqueGenre,
    title: capitalize(uniqueGenre),
  }));

  return localUniqueGenres;
};

const noop = () => {};

export {createVideoMock, adapterToLocalFilms, extend, adapterToLocalComments, GetPath, getDateFormat, getWatchTimeFormat, getUniqueGenres, noop};
