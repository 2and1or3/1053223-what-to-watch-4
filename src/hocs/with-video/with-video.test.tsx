import * as React from "react";
import * as renderer from "react-test-renderer";

import {createVideoMock} from '../../utils';
import withVideo from './with-video';
import {FilmType} from '../../types';

const film: FilmType = {
  id: `0`,
  title: `Fantastic Beasts: The Crimes of Grindelwald`,
  poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  background: `img/bg-the-grand-budapest-hotel.jpg`,
  backgroundColor: `#000`,
  cover: `img/the-grand-budapest-hotel-poster.jpg`,
  isFavorite: false,
  src: `path`,
  genre: `Drama`,
  release: `2014`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
  rating: 8.9,
  voiceCount: 240,
  duration: 99,
  director: `Wes Andreson`,
  actors: [
    `Bill Murray`,
    `Edward Norton`,
    `Jude Law`,
    `Willem Dafoe`,
    `Saoirse Ronan`,
    `Tony Revoloru`,
    `Tilda Swinton`,
    `Tom Wilkinson`,
    `Owen Wilkinson`,
    `Adrien Brody`,
    `Ralph Fiennes`,
    `Jeff Goldblum`],
};

interface MockProps {
  children: React.ReactNode;
}

const MockComponent: React.FunctionComponent<MockProps> = (props: MockProps) => {
  const {children} = props;
  return (
    <div>
      {children}
    </div>
  );
};

const MockComponentWithVideo = withVideo(MockComponent);

it(`Render Component with video`, () => {
  const tree = renderer
  .create(
      <MockComponentWithVideo
        currentFilm = {film}
        isMuted = {true}
        isPlaying = {false}
      />, {
        createNodeMock: createVideoMock
      })
  .toJSON();

  expect(tree).toMatchSnapshot();
});
