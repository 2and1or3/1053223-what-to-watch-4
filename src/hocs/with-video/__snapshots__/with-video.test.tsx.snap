// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Component with video 1`] = `
<div>
  <video
    className="player__video"
    height="175"
    width="280"
  />
</div>
`;
